### ISSUE TITLE
Parallel computing hung when using Mortar Constraint

### ISSUE THREAD

**Issue by BoZeng1997** [idaholab/moose#23796]
## Bug Description

When using Mortar method in `Constraints` of type `ModularGapConductanceConstraint` to model gap heat transfer, calculation hung at `Computing Residual`. The discussion about this issue is at https://github.com/idaholab/moose/discussions/23629#discussion-4934340.  When running on 2 nodes with 4 cpus on each node with the input file and msh mesh provided, the output of Unique Stack Traces from `find_hung_process.sh` began to diverged from `-s 35`. Here is the console output:
```
Unique Stack Traces
********************************************************************************
Count: 7

Host: dcc-core-01 PID: 1161914
Host: dcc-core-01 PID: 1161912
Host: dcc-core-01 PID: 1161915
Host: dcc-core-02 PID: 3391780
Host: dcc-core-01 PID: 1161913
Host: dcc-core-02 PID: 3391778
Host: dcc-core-02 PID: 3391777

#0  0x00007f87f10c12c8 in poll () from /lib64/libc.so.6
#1  0x00007f87f19745b3 in MPID_nem_tcp_connpoll () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#2  0x00007f87f19648b8 in MPIDI_CH3I_Progress () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#3  0x00007f87f1954772 in MPID_Iprobe () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#4  0x00007f87f17fe1e9 in PMPI_Iprobe () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#5  0x00007f87fb95a16d in Moose::Mortar::Contact::communicateGaps(std::unordered_map<libMesh::DofObject const*, std::pair<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<900ul> >, true>, double>, std::hash<libMesh::DofObject const*>, std::equal_to<libMesh::DofObject const*>, std::allocator<std::pair<libMesh::DofObject const* const, std::pair<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<900ul> >, true>, double> > > >&, unsigned int, MooseMesh const&, bool, bool, libMesh::Parallel::Communicator const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/modules/contact/lib/libcontact-opt.so.0
#6  0x00007f87fb95c417 in ComputeWeightedGapLMMechanicalContact::post() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/modules/contact/lib/libcontact-opt.so.0
#7  0x00007f87f90897f6 in ComputeMortarFunctor::operator()(Moose::ComputeType) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#8  0x00007f87f92e85fd in NonlinearSystemBase::mortarConstraints(Moose::ComputeType) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#9  0x00007f87f93c5534 in NonlinearSystemBase::computeResidualInternal(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#10 0x00007f87f93c6468 in NonlinearSystemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#11 0x00007f87f994b698 in FEProblemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#12 0x00007f87f989429a in FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#13 0x00007f87f987c6a5 in FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#14 0x00007f87f93c0542 in ComputeResidualFunctor::residual(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, libMesh::NonlinearImplicitSystem&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#15 0x00007f87f7ea467e in libmesh_petsc_snes_residual () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#16 0x00007f87f4652d2d in SNESComputeFunction () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#17 0x00007f87f46449d8 in SNESLineSearchApply_Basic () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#18 0x00007f87f464216c in SNESLineSearchApply () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#19 0x00007f87f46b3042 in SNESSolve_NEWTONLS () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#20 0x00007f87f46619b9 in SNESSolve () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#21 0x00007f87f7ea8ca0 in libMesh::PetscNonlinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#22 0x00007f87f7f22c88 in libMesh::NonlinearImplicitSystem::solve() () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#23 0x00007f87f9d5b4ab in TimeIntegrator::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#24 0x00007f87f93c22d3 in NonlinearSystem::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#25 0x00007f87f98841d6 in FEProblemBase::solve(unsigned int) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#26 0x00007f87f9f31344 in FEProblemSolve::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#27 0x00007f87f9f319c5 in FixedPointSolve::solveStep(double&, double&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#28 0x00007f87f9f48044 in FixedPointSolve::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#29 0x00007f87f97074ce in TimeStepper::step() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#30 0x00007f87f9f3052e in Transient::takeStep(double) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#31 0x00007f87f9f2d3d0 in Transient::execute() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#32 0x00007f87fa29386b in MooseApp::executeExecutioner() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#33 0x00007f87fa29dfc9 in MooseApp::run() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#34 0x0000555b1603d136 in main ()

********************************************************************************
Count: 1

Host: dcc-core-02 PID: 3391779

Host: dcc-core-02 PID: 3391779
#0  0x00007f94939192a5 in MPIDI_CH3I_Progress () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#1  0x00007f94938a0e8b in MPIR_Waitall_state () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#2  0x00007f94938a0f19 in MPIR_Waitall_impl () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#3  0x00007f94938a11a6 in MPIR_Waitall () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#4  0x00007f94937c80ca in PMPI_Waitall () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libmpi.so.12
#5  0x00007f9495c35472 in PetscSFLinkWaitRequests_MPI () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#6  0x00007f9495b993e6 in PetscSFBcastEnd_Basic () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#7  0x00007f9495b8b86f in PetscSFBcastEnd () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#8  0x00007f9495b925f6 in VecScatterEnd_Internal () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#9  0x00007f9495b97df4 in VecScatterEnd () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#10 0x00007f9495b6227b in VecGhostUpdateEnd () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#11 0x00007f9499ca1c00 in libMesh::PetscVector<double>::operator=(libMesh::PetscVector<double> const&) () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#12 0x00007f9499ca1c8a in libMesh::PetscVector<double>::operator=(libMesh::NumericVector<double> const&) () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#13 0x00007f9499e57f87 in libmesh_petsc_snes_residual () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#14 0x00007f9496606d2d in SNESComputeFunction () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#15 0x00007f94965f89d8 in SNESLineSearchApply_Basic () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#16 0x00007f94965f616c in SNESLineSearchApply () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#17 0x00007f9496667042 in SNESSolve_NEWTONLS () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#18 0x00007f94966159b9 in SNESSolve () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/lib/libpetsc.so.3.16
#19 0x00007f9499e5cca0 in libMesh::PetscNonlinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int) () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#20 0x00007f9499ed6c88 in libMesh::NonlinearImplicitSystem::solve() () from /hpc/group/dolbowlab/bz75/moose-compilers/mambaforge3/envs/moose/libmesh/lib/libmesh_opt.so.0
#21 0x00007f949bd0f4ab in TimeIntegrator::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#22 0x00007f949b3762d3 in NonlinearSystem::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#23 0x00007f949b8381d6 in FEProblemBase::solve(unsigned int) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#24 0x00007f949bee5344 in FEProblemSolve::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#25 0x00007f949bee59c5 in FixedPointSolve::solveStep(double&, double&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#26 0x00007f949befc044 in FixedPointSolve::solve() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#27 0x00007f949b6bb4ce in TimeStepper::step() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#28 0x00007f949bee452e in Transient::takeStep(double) () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#29 0x00007f949bee13d0 in Transient::execute() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#30 0x00007f949c24786b in MooseApp::executeExecutioner() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#31 0x00007f949c251fc9 in MooseApp::run() () from /hpc/group/dolbowlab/bz75/projects/raccoon/moose/framework/libmoose-opt.so.0
#32 0x000056402f39f136 in main ()
```
## Steps to Reproduce

Run the input file with the msh file in the link on multiple cpus (preferably multiple nodes). I had hard time renaming boundaries on two annular meshes generated from `AnnularMeshGenerator` so all I can offer is `msh`. I was not able to reproduce the bug on plane type gap. I guess moose generated mesh of cylinder type gap might also lead to this bug.
https://duke.box.com/s/ujb1pbjo7hpgx8lta9747w4nhk5pjpub
```
stress_free_temperature = 300
thermal_expansion_coeff = 6.66e-6

[Problem]
  type = FEProblem
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  temperature = T_K
[]

[Mesh]
  patch_update_strategy = iteration
  use_displaced_mesh = true
  patch_size = 40
  [ori]
    type = FileMeshGenerator
    file = 'pellets_transf_h0.2e-3.msh'
  []
[]

[Variables]
  [disp_x]
    block = 'pellet_inner pellet_outer '
  []
  [disp_y]
    block = 'pellet_inner pellet_outer '
  []
  [disp_z]
    block = 'pellet_inner pellet_outer '
  []
  [T_K]
    [InitialCondition]
      type = ConstantIC
      value = 300.0
    []
  []
  [lm_pellet]
    block = 'pellet_secondary_subdomain'
    use_dual = true
  []
[]

[Kernels]
  [solid_x]
    type = ADStressDivergenceTensors
    variable = disp_x
    component = 0
    block = 'pellet_inner pellet_outer'
    use_displaced_mesh = false
  []
  [solid_y]
    type = ADStressDivergenceTensors
    variable = disp_y
    component = 1
    block = 'pellet_inner pellet_outer '
    use_displaced_mesh = false
  []
  [solid_z]
    type = ADStressDivergenceTensors
    variable = disp_z
    component = 2
    block = 'pellet_inner pellet_outer '
    use_displaced_mesh = false
  []
  [timeder]
    type = ADHeatConductionTimeDerivative
    variable = 'T_K'
    density_name = density
    specific_heat = specific_heat
    block = 'pellet_inner pellet_outer '
    use_displaced_mesh = true
  []
  [diff]
    type = ADHeatConduction
    variable = 'T_K'
    thermal_conductivity = thermal_conductivity
    block = 'pellet_inner pellet_outer '
    use_displaced_mesh = true
  []
  [heatsource]
    type = ADMatHeatSource
    variable = 'T_K'
    material_property = radial_source
    block = 'pellet_inner pellet_outer '
    use_displaced_mesh = true
  []
[]


[Debug]
  show_var_residual_norms = TRUE
[]


[BCs]
  [mirror_z]
    type = ADDirichletBC
    variable = disp_z
    boundary = 'mirror_innerp mirror_outerp  '
    value = 0
  []
  [mirror_x]
    type = ADDirichletBC
    variable = disp_x
    boundary = 'mirror_innerp mirror_outerp  '
    value = 0
  []
  [mirror_y]
    type = ADDirichletBC
    variable = disp_y
    boundary = 'mirror_innerp mirror_outerp  '
    value = 0
  []
[]


[Materials]
  [pellet_properties]
    type = ADGenericConstantMaterial
    prop_names = 'density  thermal_conductivity specific_heat'
    prop_values = '3.3112e3  34 1.2217e3' 
    block = 'pellet_inner pellet_outer'
  []
  [pulse_shape_linear]
    type = ADGenericFunctionMaterial
    prop_values = '5e10*max(11455*(t)/7,1e-9)'
    prop_names = 'radial_source'
    output_properties = 'radial_source'
    block = 'pellet_inner pellet_outer'
    use_displaced_mesh = false
  []
  [strain]
    type = ADComputeSmallStrain
    displacements = 'disp_x disp_y disp_z'
    eigenstrain_names = eigenstrain #nameS!
    block = 'pellet_inner pellet_outer '
  []
  [thermal_strain]
    type = ADComputeThermalExpansionEigenstrain
    stress_free_temperature = ${stress_free_temperature} 
    thermal_expansion_coeff = ${thermal_expansion_coeff} 
    eigenstrain_name = eigenstrain
    block = 'pellet_inner pellet_outer '
  []
  [elasticity]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 3.306e11
    poissons_ratio = 0.329
  []
  [stress]
    type = ADComputeLinearElasticStress
    block = 'pellet_inner pellet_outer '
  []
[]



[Contact]
  [pellet]
    primary = void_pellet_0
    secondary = void_pellet_1
    model = frictionless
    penalty = 1e6
    formulation = mortar
    c_normal = 1e6 
    correct_edge_dropping = true
  []
[]


[UserObjects]
  [conduction]
    type = GapFluxModelConduction
    temperature = T_K
    boundary = 'void_pellet_0 void_pellet_1    '
    gap_conductivity = 0.4
    use_displaced_mesh = true
  []
  [rad_pellet]
    type = GapFluxModelRadiation
    temperature = T_K
    boundary = void_pellet_0
    primary_emissivity = 0.37
    secondary_emissivity = 0.37
    use_displaced_mesh = true
  []
[]

[Constraints]
  [gap_pellet]
    type = ModularGapConductanceConstraint
    variable = lm_pellet
    secondary_variable = T_K
    primary_boundary = 'void_pellet_0'
    primary_subdomain = pellet_primary_subdomain
    secondary_boundary = 'void_pellet_1'
    secondary_subdomain = pellet_secondary_subdomain
    gap_flux_models = 'conduction rad_pellet' #closed_pellet  
    gap_geometry_type = 'CYLINDER'
    cylinder_axis_point_1 = '0 0 0'
    cylinder_axis_point_2 = '0 0 1'
    use_displaced_mesh = true
    quadrature = SECOND
  []
[]


[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -pc_factor_shift_type -pc_factor_shift_amount '
  petsc_options_value = 'lu      superlu_dist NONZERO 1e-14' 
  automatic_scaling = true
  line_search = none
  ignore_variables_for_autoscaling = 'pellet_normal_lm  lm_pellet '
  compute_scaling_once = true 
  scaling_group_variables = 'disp_x disp_y disp_z; T_K' 
  nl_rel_tol = 1e-50
  nl_abs_tol = 1e-8
  nl_max_its = 20
  dtmin = 1e-3
  dt = 1e-3
  start_time = 0e-3
  end_time = 7e-3
[]

[Outputs]
  [exodus]
    type = Exodus
    file_base = constMat
  []
  print_linear_residuals = false
[]
```

**Comment by lindsayad** [comment]
@recuero haven't you said you've gotten hangs in the past with mortar?

**Comment by recuero** [comment]
It's been a while since I've got a hang with mortar. But, yes, I have. This is a 3D problem, which has been used less than 2D.

**Comment by lindsayad** [comment]
I usually configure with size 300 to be safe ... I know the number can get pretty high for 3D mortar

**Comment by BoZeng1997** [comment]
@roystgnr I set `--with-derivative-size=700`

**Comment by roystgnr** [comment]
Wow that's huge.  I seem to be able to replicate the issue with 300, though, so I'll stick with that unless I run into evidence that it's a problem later.

**Comment by roystgnr** [comment]
Heh, typo; that should have been "rank 6".  If I ever see an error from rank 8 on a case where 0-7 are running then we're *really* in deep.

**Comment by roystgnr** [comment]
Throwing more `parallel_object_only()` and `libmesh_parallel_only(communicator)` calls into MOOSE is at least suggesting that the NBX stuff is a red herring; that's just the first time the Mortar code gets back into a libMesh or TIMPI call which does that check (or which blocks on MPI, in opt mode).

At the moment in `devel` runs I'm seeing 7 ranks entering `Moose::Mortar::Contact::communicateGaps` at the same time one rank is entering `FEProblemBase::computingNonlinearResid`.  It does still look like the divergence is in `libmesh_petsc_snes_residual`, though, which I have no explanation for.  I probably ought to go back to `dbg` since the inlining in `devel` makes the stacks slightly more confusing.

About to take a break for dinner; I might have time to look at this again late tonight, but figured I'd throw what I have into the thread in case it jogs ideas for anyone.

**Comment by roystgnr** [comment]
Ok, dinner slightly delayed.  In `dbg` it looks like everybody's making it into `ComputeResidualFunctor::residual`, but from there we have most processors in `FEProblemBase::computeResidualSys` and one in `FEProblemBase::computingNonlinearResid`

**Comment by lindsayad** [comment]
The latter is just a three-liner boolean setter
```c++
void
FEProblemBase::computingNonlinearResid(bool computing_nonlinear_residual)
{
  if (_displaced_problem)
    _displaced_problem->computingNonlinearResid(computing_nonlinear_residual);
  _computing_nonlinear_residual = computing_nonlinear_residual;
}
```

**Comment by roystgnr** [comment]
Egg on my face; a `parallel_object_only()` doesn't belong in a boolean setter.  Seems to have been informative in this case, though.  I wonder how that rank is making it there ahead of the others.  I'd have guessed there might be an exception thrown that is getting a serial rather than a parallel catch ... except that I have `catch throw` on by default in my gdb config and I'm not seeing anything.

**Comment by lindsayad** [comment]
Hmm an exception thrown on just one process really does seem like a tantalizing explanation though

**Comment by lindsayad** [comment]
Are you seeing exceptions thrown at any point? Our parser relies heavily on throwing and catching exceptions so you should be seeing thrown exceptions for every MOOSE simulation you run if you have `catch throw` enabled

**Comment by lindsayad** [comment]
hmm maybe something changed? I just ran an input and indeed didn't see any thrown exceptions during parsing

**Comment by roystgnr** [comment]
Ah, shoot.  I had `catch throw` turned off again in .gdbinit; at some point I decided that stopping at `report_error` in libMesh and TIMPI was enough.

The problem is definitely an exception thrown on one processor.  `mooseException("Quadrature point: ", qp_back, " out of bounds, truncating");` in MortarUtils.C

**Comment by roystgnr** [comment]
Everybody makes it to `NonlinearSystemBase::mortarConstraints`, then almost everybody continues, but one processor is throwing that exception in `ComputeMortarFunctor::operator()` -> `(some lambda function)` -> Moose::Mortar::loopOverMortarSegments` -> `Moose::Mortar::projectQPoints3d`.  Its stack then unwinds all the way back to `NonlinearSystemBase::computeResidualTags`, but the rest of the ranks still have parallel things to do later via other `mortarConstraints` calls.

We ought to be able to catch deeper in the stack, sync up in parallel, and re-throw.  Let me see if I can fix this.

**Comment by roystgnr** [comment]
`PARALLEL_CATCH` isn't the right thing to do for this, is it?  Or is `checkExceptionAndStopSolve` more recoverable and less ominous than it sounds?

**Comment by roystgnr** [comment]
And now I get to:

```
Quadrature point: (x,y,z)=(0.000311332,       -1,        0) out of bounds, truncating
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge


*** ERROR ***
The following error occurred in the object "TimeStepper", of type "ConstantDT".

Solve failed and timestep already at or below dtmin, cannot continue!
```

Which I'm guessing is a whole other problem we need to fix (doesn't dtmin default to like 1e-12? we're at 1e-3!), but at least the desynchronization problem seems to be solved.  I'll put up a PR to close this shortly, and if there's something deeper with the convergence issues that can be a separate ticket.

**Comment by lindsayad** [comment]
In the input @BoZeng1997 sent us `dtmin` is set to the same value as `dt`. That's why you're seeing the `cannot continue` message
